pkg/lib/catalogsource/image_template.go

```diff
@@ -0,0 +1,356 @@
+				foundEntry := false
+				for _, existingJsonPath := range existingJsonPaths {
+					if jsonPath == existingJsonPath {
+						foundEntry = true
+					}
+				}
+				// if we did not find a jsonpath entry then add it now
+				if !foundEntry {
+					gvkToJSONPathMap[key] = append(gvkToJSONPathMap[key], jsonPath)
+				}
+			} else {
+				gvkToJSONPathMap[key] = append(gvkToJSONPathMap[key], jsonPath)
+			}
+
+		}
+	}
+	return foundGVKs
+}
+
+func UpdateGVKValue(u *unstructured.Unstructured, logger *logrus.Logger) {
+
+	// reconstitute the key
+	key := GVK_Key{
+		GroupVersionKind: u.GetObjectKind().GroupVersionKind(),
+		name:             u.GetName(),
+		namespace:        u.GetNamespace(),
+	}
+
+	// find the JSON paths
+	if jsonPaths, ok := gvkToJSONPathMap[key]; ok {
+
+		// convert the unstructured object into JSON as bytes
+		jsonBytes, err := u.MarshalJSON()
+		if err != nil {
+			logger.WithError(err).Warn("unable to convert kube manifest to JSON")
+		}
+
+		// pass the JSON as bytes through the go json library so its in the right format
+		var processedJSON interface{}
+		err = json.Unmarshal(jsonBytes, &processedJSON)
+		if err != nil {
+			logger.WithError(err).Warn("unable to convert kube manifest json data into usable form")
+		}
+
+		for _, jsonPath := range jsonPaths {
+
+			gvkLogger := logger.WithFields(logrus.Fields{
+				"gvk":      u.GetObjectKind().GroupVersionKind().String(),
+				"jsonPath": jsonPath,
+			})
+			// create the json path parser
+			jsonPathParser := jsonpath.New("GVK path parser")
+
+			// parse the json path template
+			err = jsonPathParser.Parse(jsonPath)
+			if err != nil {
+				gvkLogger.WithError(err).Warn("unable to parse json path template")
+				continue
+			}
+
+			// execute the parser using the JSON data writing the results into a buffer
+			buf := new(bytes.Buffer)
+			err = jsonPathParser.Execute(buf, processedJSON)
+			if err != nil {
+				gvkLogger.WithError(err).Warn("unable to execute json path parsing")
+				continue
+			}
+
+			templateMapKey := convertToKey(key, jsonPath)
+			templateMapValue := buf.String()
+
+			if jsonPath == templateMapValue {
+				// the jsonpath is exactly the same as the templateMapValue, this means
+				// that the jsonpath was probably invalid, so don't update
+				gvkLogger.Debugf("jsonpath %q is likely invalid (maybe curly braces are missing?)", jsonPath)
+				continue
+			}
+			// reconstruct the key for the template replacement map and add
+			// whatever we got from the json path execution
+			templateNameToReplacementValuesMap[templateMapKey] = templateMapValue
+			gvkLogger.Debugf("updated templateNameToReplacementValuesMap: key=%q value%q", templateMapKey, templateMapValue)
+		}
+	}
+}
+
+func UpdateKubeVersion(serverVersion *version.Info, logger *logrus.Logger) {
+	if serverVersion == nil {
+		logger.Warn("no server version provided")
+		return
+	}
+
+	templateNameToReplacementValuesMap[TEMPL_KUBEMAJORV] = serverVersion.Major
```

Comment:
Not threadsafe -- I suggest using a mutex here.

Reply:
fixed